homeassistant/components/device_tracker/automatic.py

```diff
@@ -38,65 +44,183 @@
-                         "location:updated events will not be received.")
-            session = yield from client.create_session_from_password(
-                DEFAULT_SCOPE, config[CONF_USERNAME], config[CONF_PASSWORD])
 
+    filename = AUTOMATIC_CONFIG_FILE.format(config[CONF_CLIENT_ID])
+    refresh_token = yield from hass.async_add_job(
+        _get_refresh_token_from_file, hass, filename)
+
+    @asyncio.coroutine
+    def initialize_data(session):
+        """Initialize the AutomaticData object from the created session."""
+        hass.async_add_job(
+            _write_refresh_token_to_file, hass, filename,
+            session.refresh_token)
         data = AutomaticData(
             hass, client, session, config[CONF_DEVICES], async_see)
 
         # Load the initial vehicle data
         vehicles = yield from session.get_vehicles()
         for vehicle in vehicles:
             hass.async_add_job(data.load_vehicle(vehicle))
-    except aioautomatic.exceptions.AutomaticError as err:
-        _LOGGER.error(str(err))
-        return False
 
-    @callback
-    def ws_connect(event):
-        """Open the websocket connection."""
-        hass.async_add_job(data.ws_connect())
+        # Create a task instead of adding a tracking job, since this task will
+        # run until the websocket connection is closed.
+        hass.loop.create_task(data.ws_connect())
 
-    @callback
-    def ws_close(event):
-        """Close the websocket connection."""
-        hass.async_add_job(data.ws_close())
+    if refresh_token is not None:
+        try:
+            session = yield from client.create_session_from_refresh_token(
+                refresh_token)
+            yield from initialize_data(session)
+            return True
+        except aioautomatic.exceptions.BadRequestError as err:
+            if str(err) == 'err_invalid_refresh_token':
+                _LOGGER.error("Stored refresh token was invalid.")
+                yield from hass.async_add_job(
+                    _write_refresh_token_to_file, hass, filename, None)
+            else:
+                _LOGGER.error(str(err))
+                return False
+        except aioautomatic.exceptions.AutomaticError as err:
+            _LOGGER.error(str(err))
+            return False
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, ws_connect)
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, ws_close)
+    if CONF_USERNAME in config:
+        try:
+            session = yield from client.create_session_from_password(
+                scope, config[CONF_USERNAME], config[CONF_PASSWORD])
+            yield from initialize_data(session)
+            return True
+        except aioautomatic.exceptions.AutomaticError as err:
+            _LOGGER.error(str(err))
+            return False
+
+    configurator = get_component('configurator')
+    request_id = configurator.async_request_config(
+        hass, "Automatic", lambda _: None,
```

Comment:
Oh I get it, automatic will hit the callback API.

Should we make callback optional for configurator?